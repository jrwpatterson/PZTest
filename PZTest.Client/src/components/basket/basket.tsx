import React from 'react'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import Divider from '@material-ui/core/Divider'

import { useBasketHook } from '../../hooks'
import { Button } from '@material-ui/core'

interface BasketProps {}

export const Basket = (props: BasketProps) => {
  const { basket, clearBasket, removeProductFromBasket } = useBasketHook()
  return (
    <div>
      <List>
        <ListItem>
          <ListItemText primary='Basket' />
        </ListItem>
        <Divider />
        {basket.lines?.map(line => {
          return (
            <ListItem key={`${(line.productID, line.rowNo)}`}>
              <ListItemText primary={line.name} />
              <ListItemText primary=' - ' />
              <ListItemText primary={`$${line.price ?? 0}`} />
            </ListItem>
          )
        })}
        <Divider />
        <ListItem data-testid='basket.total'>
          <ListItemText primary='Total' />
          <ListItemText primary=' - ' />
          <ListItemText
            data-testid='basket.total'
            primary={`$${basket.total ?? 0}`}
          />
        </ListItem>

        <Divider />
        <Button variant='outlined' color='secondary' onClick={clearBasket}>
          Clear Basket
        </Button>
      </List>
    </div>
  )
}
