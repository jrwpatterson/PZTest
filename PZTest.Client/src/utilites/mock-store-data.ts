import { RootReducer } from '../root-reducer'

export const defaultMockStore: RootReducer = {
  basketReducer: {
    basket: {
      id: 'testID',
      lines: [
        {
          grams: 200,
          name: 'test product',
          productID: 'testProdID',
          rowNo: 0,
          price: 5,
        },
        {
          grams: 400,
          name: 'test product 2',
          productID: 'testProdID2',
          rowNo: 1,
          price: 20,
        },
      ],
      total: 25,
    },
    validBasket: true,
  },
  cheeseReducer: {
    cheeses: [
      {
        cheeseColour: 'purple',
        id: 'testid',
        name: 'testCheese',
        pictureUrl: 'testUrl',
        pricePerKG: 10,
      },
      {
        cheeseColour: 'pink',
        id: 'testid2',
        name: 'testPinkCheese',
        pictureUrl: 'testPinkUrl',
        pricePerKG: 15,
      },
    ],
  },
}
