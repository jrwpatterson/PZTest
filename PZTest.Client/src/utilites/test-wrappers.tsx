import React, { ReactType, ReactNode } from 'react'
import { Provider } from 'react-redux'
import configureStore, { MockStoreEnhanced } from 'redux-mock-store'
import { defaultMockStore } from './mock-store-data'
import { render } from '@testing-library/react'
import { renderHook, RenderHookOptions } from '@testing-library/react-hooks'
import { RootReducer } from '../root-reducer'
// import { createMount } from '@material-ui/core/test-utils'
import { Provider as FetchProvider } from 'use-http'
import { MemoryRouter } from 'react-router-dom'

const mockStore = configureStore([])

const createStore = (storeData: RootReducer) => {
  return mockStore(storeData)
}

export const createMockProvider = (
  storeData: RootReducer = defaultMockStore,
) => {
  return ({ children }: { children?: ReactNode }) => (
    <FetchProvider url='http://localhost:5001/'>
      <Provider store={createStore(storeData)}>{children}</Provider>
    </FetchProvider>
  )
}

export const createMockProviderWithStore = (
  fakeStore: MockStoreEnhanced<unknown, {}>,
) => {
  return ({ children }: { children?: ReactNode }) => (
    <FetchProvider url='http://localhost:5001/'>
      <Provider store={fakeStore}>{children}</Provider>
    </FetchProvider>
  )
}

export const renderReduxConnectedComponent = (
  children: ReactType | JSX.Element,
  storeOverride: RootReducer = defaultMockStore,
  initialEntries?: string[],
) => {
  const storeData = storeOverride || defaultMockStore
  // const mount = createMount()
  const store = createStore(storeData)
  const Wrapper = createMockProviderWithStore(store)
  const element = (
    <MemoryRouter initialEntries={initialEntries}>
      <Wrapper>{children}</Wrapper>
    </MemoryRouter>
  )
  const component = render(element)

  return {
    store,
    ...component,
  }
}

export const renderReduxConnectedHook = <T extends {}>(
  callback: (props: unknown) => T,
  storeData: RootReducer = defaultMockStore,
  options: RenderHookOptions<unknown> = {},
) => {
  const store = createStore(storeData)
  const wrapper = createMockProviderWithStore(store)
  const renderedHook = renderHook(callback, { ...options, wrapper })

  return {
    store,
    ...renderedHook,
  }
}
