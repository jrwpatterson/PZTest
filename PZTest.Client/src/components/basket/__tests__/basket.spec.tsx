import React from 'react'
import { Basket } from '../basket'
import { renderReduxConnectedComponent } from '../../../utilites/test-wrappers'

describe('basket', () => {
  it('should match its snapshot', () => {
    const { container } = renderReduxConnectedComponent(<Basket />)
    expect(container).toMatchSnapshot()
  })

  it('should show the basket total', () => {
    const { container } = renderReduxConnectedComponent(<Basket />)
  })

  it('should show the lines', () => {
    throw new Error('not implemented')
  })

  it('should have a clear button that clears all', () => {
    throw new Error('not implemented')
  })

  it('it should have a remove line for each line', () => {
    throw new Error('not implemented')
  })
})
