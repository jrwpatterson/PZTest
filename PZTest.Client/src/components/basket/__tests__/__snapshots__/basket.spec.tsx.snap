// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basket should match its snapshot 1`] = `
<div>
  <div>
    <ul
      class="MuiList-root MuiList-padding"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Basket
          </span>
        </div>
      </li>
      <hr
        class="MuiDivider-root"
      />
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            test product
          </span>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
             - 
          </span>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            $5
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            test product 2
          </span>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
             - 
          </span>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            $20
          </span>
        </div>
      </li>
      <hr
        class="MuiDivider-root"
      />
      <li
        class="MuiListItem-root MuiListItem-gutters"
        data-testid="basket.total"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            Total
          </span>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
             - 
          </span>
        </div>
        <div
          class="MuiListItemText-root"
          data-testid="basket.total"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
          >
            $25
          </span>
        </div>
      </li>
      <hr
        class="MuiDivider-root"
      />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Clear Basket
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </ul>
  </div>
</div>
`;
